### ISSUE TITLE
allow comptime unreachable prong for errors not in error set

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#22138]
```zig
const Error = error{ A, B };

test "example" {
    foo(1) catch |err| switch (err) {
        error.A => {},
        error.B => {},
        error.C => comptime unreachable,
    };
}

fn foo(x: i32) !void {
    switch (x) {
        0 => return error.A,
        1 => return error.B,
        else => return,
    }
}
```

Current behavior is a compile error:

```
test.zig:7:9: error: expected type '@typeInfo(@typeInfo(@TypeOf(test.foo)).@"fn".return_type.?).error_union.error_set', found 'error{C}'
        error.C => comptime unreachable,
        ^~~~~~~
test.zig:7:9: note: 'error.C' not a member of destination error set
```

I propose instead this construct be allowed, since receiving `error.C` is indeed comptime unreachable exactly due to not being in the error set, and this provides a way to assert (at compile-time) that no such error is ever added to the set.

This pattern is especially important when using the common pattern of having a particular error in a set indicate that an error message was already created and stored in a separate location.

Currently I am dealing with a large refactor that is being hamstrung by lack of this functionality.

**Comment by mlugg** [comment]
On a related note, we currently allow a few forms of redundant `else =>` prong for switches on errors. However, the way this is currently implemented is incorrect. It's done by this logic in Sema:

https://github.com/ziglang/zig/blob/6188cb8e50882a55c90578114556c6d1b7073e9d/src/Sema.zig#L14066-L14105

Reproduced here since GitHub refuses to embed code references into `Sema.zig`:
```zig
            if (has_else and seen_errors.count() == error_names.len) {
                // In order to enable common patterns for generic code allow simple else bodies
                // else => unreachable,
                // else => return,
                // else => |e| return e,
                // even if all the possible errors were already handled.
                const tags = sema.code.instructions.items(.tag);
                const datas = sema.code.instructions.items(.data);
                for (else_case.body) |else_inst| switch (tags[@intFromEnum(else_inst)]) {
                    .dbg_stmt,
                    .dbg_var_val,
                    .ret_type,
                    .as_node,
                    .ret_node,
                    .@"unreachable",
                    .@"defer",
                    .defer_err_code,
                    .err_union_code,
                    .ret_err_value_code,
                    .save_err_ret_index,
                    .restore_err_ret_index_unconditional,
                    .restore_err_ret_index_fn_entry,
                    .is_non_err,
                    .ret_is_non_err,
                    .condbr,
                    => {},
                    .extended => switch (datas[@intFromEnum(else_inst)].extended.opcode) {
                        .restore_err_ret_index => {},
                        else => break,
                    },
                    else => break,
                } else break :else_validation;


                return sema.fail(
                    block,
                    else_case.src,
                    "unreachable else prong; all cases already handled",
                    .{},
                );
            }
```

This is not okay: we can't just try to match patterns in ZIR. Doing this will essentially always leak compiler implementation details; for instance, here's a case I found after a quick minute of playing around:
```zig
export fn foo() u32 {
    switch (error.Something) {
        error.Something => {},
        else => return 2,   // this triggers a compile error...
        //else => return 0, // ...but this works fine due to ZIR `int(...)` elision for zero
    }
    return 1;
}
```

So, when we implement this proposal, we should perform the relevant check in AstGen and represent it in the ZIR, and we should rework this existing check to do the same.

**Comment by nektro** [comment]
strongly hope this does not make it in because it would be another way for dead code to sneak its way into codebases with no easy way to detect it. using compile error-guided adding and deleting is one of my favorite reasons to use `switch`. here being the use case of an error is removed from an explicit or inferred error set and zig kindly shows you all the cases to remove.

additionally,

```zig
const E = enum { a, b };
var instance: E = .a;
switch (instance) {
  .a => {},
  .b => {},
  // with this proposal it would make sense for this to compile but that feels quite silly
  .c => comptime unreachable,
}
```


> Currently I am dealing with a large refactor that is being hamstrung by lack of this functionality.

very curious if you could elaborate on this as it seems to be the root of the inspiration for this proposal

**Comment by chadwain** [comment]
This is similar to #21885

**Comment by scottredig** [comment]
Is there a strong reason to do this as a language change, when this is doable in user code? eg:

```zig
const std = @import("std");

test "example" {
    foo(1) catch |err| {
        forbidErrors(err, error.C);
        switch (err) {
            error.A => {},
            error.B => {},
        }
    };
}

fn foo(x: i32) !void {
    switch (x) {
        0 => return error.A,
        1 => return error.B,
        // 2 => return error.C,
        else => return,
    }
}

inline fn forbidErrors(err: anytype, comptime forbidden_errors: anytype) void {
    switch (err) {
        inline else => |have_error| switch (forbidden_errors) {
            inline else => |forbidden_error| if (have_error == forbidden_error) {
                @compileError("Error " ++ @errorName(forbidden_error) ++ " must not appear in " ++ @typeName(@TypeOf(err)));
            },
        },
    }
}
```

`inline` here stops the compile at the call to `forbidErrors`, instead of giving the error for a missing switch prong before also giving the `@compileError` erorr.

**Comment by rohlem** [comment]
In status-quo, `unreachable` is a value which triggers a panic when evaluated. It is put in places which you intend execution to never hit.
Because `comptime` enacts execution to fully evaluate something, `comptime unreachable` should, if it is reached in analysis, always result in a compile error right there.
I don't like overloading the current syntax with a different meaning.

Status-quo already provides all the tools for a pass-through function to check that an error union contains no errors from a given set.
`assertUnreachableErrors(x, error{A,B,C})` can assert at compile time that `@typeInfo(x).ErrorUnion.error_set` contains no error of the given set, and return runtime `x` unmodified.
In context:
```zig
foo(1) catch |err| switch (assertUnreachableErrors(err, error{
        C, //unreachable because ...
        D, //must already be handled in ...
    })) {
    error.A => {},
    error.B => {},
};
```

**Comment by rohlem** [comment]
> > I don't like overloading the current syntax with special-casing for a different effect here.
> 
> I'm a little confused by this point. \[...\] The behaviour of the actual `comptime unreachable` expression is entirely unchanged.

Current semantics say that as soon as analysis hits `comptime unreachable`, a compile error is triggered.
The proposal is, if a control flow construct should result in a type error (in this case a prong doesn't fit the subject's type), look if the prong body matches this particular compile-error construct.

Maybe comparing whether the prong body exactly matches `comptime unreachable` doesn't constitute analysis.
(My simple intuition is: If the compiler "looks at" `comptime unreachable`, I expect a compile error.)
Either way, there are other compile error constructs (such as `@compileError`).
Special-casing this particular one seems arbitrary, and my intuition of the behavior of `comptime unreachable`,
constructed by its parts in status-quo, does not in any way relate to `switch` prong type checks.
In short, I would see it as an unintuitive/unexpected special-case.
(At the very least it would redefine and extend the status-quo semantics of these keywords. I've provided a status-quo userland solution above, which I would prefer.)

**Comment by mlugg** [comment]
> My simple intuition would be: If the compiler "looks at" `comptime unreachable`, I expect a compile error.

This intuition simply doesn't align with how Zig works, even today. For instance, compile errors can and do occur in code which isn't semantically analyzed; these are `ast-check` errors, and are a very intentional feature of the language. Semantic analysis isn't necessary for the compiler to "look at" a piece of code.

> Under this proposal, in particular contexts it can prevent a different compile error. To me this seems more complex [...]

We already allow redundant `else => |e| return e` in switches on error sets, and I don't think this causes any confusion. `comptime unreachable` annotates a code path which will never be analyzed, and unreachable switch prongs are always either a compile error or not analyzed (for instance, the latter is the case for `noreturn` union fields). The only thing proposed here is to change redundant prongs from one of these behaviours to the other in a case where this is strictly more useful.

**Comment by rohlem** [comment]
> > My simple intuition would be: If the compiler "looks at" `comptime unreachable`, I expect a compile error.
> 
> This intuition simply doesn't align with how Zig works, even today.

For clarification, this wasn't a description how all compile errors work (I know about AST checks), just about `comptime unreachable`.
You can verify the status-quo behavior with `test {if (true) comptime unreachable;}` vs `test {if (false) comptime unreachable;}`.
(The same, logically, happens for `comptime {if (false) unreachable;}` vs `comptime {if (false) unreachable;}`.)

> > Under this proposal, in particular contexts it can prevent a different compile error. To me this seems more complex [...]
> 
> We already allow redundant `else => |e| return e` in switches on error sets, and I don't think this causes any confusion.

I agree that writing code for cases which are `unreachable` is a valuable feature (specifically for all my over-generic code, where types change between instantiations).
I often wish the compiler/language would allow me to do this more (easily).

> `comptime unreachable` annotates a code path which will never be analyzed

Again, this seems like part of the proposal.
This would be a new effect of `comptime unreachable` on the code path to get to it, which it doesn't have in status-quo.

>unreachable switch prongs are always either a compile error or not analyzed (for instance, the latter is the case for `noreturn` union fields). The only thing proposed here is to change redundant prongs from one of these behaviours to the other in a case where this is strictly more useful.

The way I read your comment is "We propose to allow redundant prongs.".
If that is the change, I am 100% in favor.
The way I read the proposal was "We special-case the existing compile error logic for prongs with bodies that exactly matches `comptime unreachable` to suppress the compile error that is otherwise always triggered.".
That seems unintuitive and overly complex to me. (And this will be my last mention of `assertUnreachableErrors` I implemented above.)

**Comment by andrewrk** [comment]
Example that already works to take advantage of `comptime unreachable`:

```zig
test "example" {
    var x: u8 = 254;
    x += 1;
    switch (x) {
        0 => {
            return;
        },
        1...254 => {
            return;
        },
        255 => {
            // oops, this was supposed to also be noreturn
        },
    }
    unreachable;
}
```

```
$ stage4/bin/zig test test.zig 
thread 520319 panic: reached unreachable code
/home/andy/src/zig/build-release/test.zig:15:5: 0x1040a46 in test.example (test)
    unreachable;
    ^
/home/andy/src/zig/lib/compiler/test_runner.zig:208:25: 0x10d9863 in mainTerminal (test)
        if (test_fn.func()) |_| {
                        ^
/home/andy/src/zig/lib/compiler/test_runner.zig:57:28: 0x10d19e0 in main (test)
        return mainTerminal();
                           ^
/home/andy/src/zig/lib/std/start.zig:608:22: 0x10d0f36 in posixCallMainAndExit (test)
            root.main();
                     ^
/home/andy/src/zig/lib/std/start.zig:248:5: 0x10d0b5f in _start (test)
    asm volatile (switch (native_arch) {
    ^
???:?:?: 0x0 in ??? (???)
error: the following test command crashed:
/home/andy/src/zig/.zig-cache/o/f84ff937b50b9aa6432d1796fc75a556/test --seed=0x68c4267c
```

Here we have a runtime safety check to make sure the unreachable is not hit. But if we make it `comptime unreachable;` then we get a handy compile error instead, ensuring that all branches are noreturn.

```
$ stage4/bin/zig test test.zig 
test.zig:15:14: error: reached unreachable code
    comptime unreachable;
             ^~~~~~~~~~~
```

And sure enough, adding the `return` into the switch prong makes the test pass:

```diff
-            // oops, this was supposed to also be noreturn
+            return;
```

```
$ stage4/bin/zig test test.zig 
All 1 tests passed.
```

This is an especially important trick when combined with the recent language enhancement that landed, [labeled switch continue](https://github.com/ziglang/zig/issues/8220).

And of course there's the infamous [errdefer comptime unreachable;](https://matklad.github.io/2024/03/21/defer-patterns.html#Statically-Enforcing-Absence-of-Errors).

**Comment by T-727** [comment]
Not sure how possible or in scope of this proposal what I'm about to suggest is, but I'd like to say that if `comptime unreachable` was also allowed on errors which **are** in the error set we could have possibilites like this:
```zig
const Mode = enum { @"async", sync };
fn open(mode: Mode) error{ AsyncError, SyncError }!void {
    ...
    switch (mode) {
        .@"async" => {
            ...
            if (...) return error.AsyncError;
        },
        .sync => {
            ...
            if (...) return error.SyncError;
        },
    }
}

fn run() !void {
    const mode: Mode = .sync;
    ...
    open(mode) catch |err| switch (err) {
        error.Sync => |e| return e,
        error.Async => comptime unreachable,
    };
}
```
Here if the programmer used regular `unreachable` assuming `mode` was a constant then later down the line it became a variable, they might forget to change that prong to handle the new usecase and live with a reachable unreachable waiting to crash.
But if it was `comptime unreachable`, and `mode` became runtime the compiler would error with:
```shell
src/main.zig:..:..: error: unable to evaluate comptime expression
```
somewhat inspired by https://github.com/ziglang/zig/issues/509

**Comment by mlugg** [comment]
That already works exactly as you describe.

**Comment by vadim-za** [comment]
> strongly hope this does not make it in because it would be another way for dead code to sneak its way into codebases with no easy way to detect it. using compile error-guided adding and deleting is one of my favorite reasons to use `switch`. here being the use case of an error is removed from an explicit or inferred error set and zig kindly shows you all the cases to remove.
> 
> additionally,
> 
> const E = enum { a, b };
> var instance: E = .a;
> switch (instance) {
>   .a => {},
>   .b => {},
>   // with this proposal it would make sense for this to compile but that feels quite silly
>   .c => comptime unreachable,
> }
> > Currently I am dealing with a large refactor that is being hamstrung by lack of this functionality.
> 
> very curious if you could elaborate on this as it seems to be the root of the inspiration for this proposal

Actually there is a strongly related case which makes lots of sense to me. Imagine the following generic function
```zig
fn f(E: type, e: E) void {
    switch (e) {
        .a => {},
        .b => {},
        else => {},
    }
}
```
It is intended and is able to handle any enum, containing fields `a`, and `b`, except when `a` and `b` are the only fields in the enum. The idea with `comptime unreachable` won't exactly help here though, but I thought I'd mention this as another related case.

Edit: the same problem exists if a switch is e.g. used to handle events published by a framework (as an enum or tagged union). A switch which _just happens_ to handle all events cannot have an else `prong` for unhandled (or default-handled, like in WinAPI) events, and will therefore break once further events are added to the framework.

**Comment by mlugg** [comment]
https://codeberg.org/ziglang/zig/pulls/30776
